Cmdline parser improvements

NightTsarina opened the pull request:
This PR changes the way the destaddr parameter is parsed, using net.HardwareAddr, which allows for different hw address syntaxes. Plus, it adds the local, all, and broadcast aliases, like the open-plc-tools package.

A second commit also adds more verbose logging to the received HP frames.

[Migrated from brandond#4]

NightTsarina committed:
Add a kingpin parser for the dest MAC address.

It also adds support for "all", "local", and "broadcast" aliases, like
open-plc-tools. And by using net.HardwareAddr, it automatically supports
different HWA formats (xx:xx, xx-xx, xxxx.xxxx) and lengths.

NightTsarina committed:
Improve logging of the status structures.

dswarbrick commented on homeplug_exporter.go:
```diff
@@ -131,6 +133,7 @@ func (n *HomeplugNetworkInfo) UnmarshalBinary(b []byte) error {
 			return err
 		}
 		n.Networks = append(n.Networks, ns)
+		log.Debugf("Network found: %s", &ns)
```
This appears to use the old logrus logging (or standard library log package), and needs to be migrated to the go-kit/log API.

NightTsarina replied:
argh, of course. This was old code and I just rebased it, sorry about that!

dswarbrick commented on homeplug_exporter.go:
```diff
@@ -143,26 +146,34 @@ func (n *HomeplugNetworkInfo) UnmarshalBinary(b []byte) error {
 			return err
 		}
 		n.Stations = append(n.Stations, ss)
+		log.Debugf("Station found: %s", &ss)
```
See previous comment about go-kit/log API.

dswarbrick commented on macaddress.go:
```diff
@@ -0,0 +1,42 @@
+	return err
+}
+
+func (f *macAddressValue) Get() interface{} { return (net.HardwareAddr)(*f.v) }
```
Style nit: my personal preference would be to expand this out over multiple lines for better readability.

NightTsarina replied:
Mine too.. I don't remmeber why I left it like this :)

dswarbrick commented on macaddress.go:
```diff
@@ -0,0 +1,42 @@
+
+func (f *macAddressValue) Get() interface{} { return (net.HardwareAddr)(*f.v) }
+
+func (f *macAddressValue) String() string { return fmt.Sprintf("%v", *f.v) }
```
Style nit: my personal preference would be to expand this out over multiple lines for better readability.

dswarbrick commented on macaddress.go:
```diff
@@ -0,0 +1,42 @@
+import (
+	"errors"
+	"fmt"
+	"gopkg.in/alecthomas/kingpin.v2"
```
Style nit: separate third-party imports from standard library imports.

NightTsarina replied:
good point

NightTsarina committed:
Fixes from code review.

NightTsarina commented:
Comments addressed. PTAL?

dswarbrick commented on macaddress.go:
```diff
@@ -0,0 +1,47 @@
+}
+
+func (f *macAddressValue) String() string {
+	return fmt.Sprintf("%v", *f.v)
```
```suggestion
	return f.v.String()
```
`fmt` import is made redundant by this also.

dswarbrick commented on macaddress.go:
```diff
@@ -0,0 +1,47 @@
+		s = "00:b0:52:00:00:01"
+	}
+	v, err := net.ParseMAC(s)
+	if err == nil && len(v) != 6 {
+		return errors.New("Invalid address length")
+	}
+	if err == nil {
+		*f.v = (net.HardwareAddr)(v)
+	}
+	return err
```
```suggestion
	if err != nil {
		return err
	} else if len(v) != 6 {
		return errors.New("invalid address length")
	}
	*f.v = v
	return nil
```

dswarbrick commented on macaddress.go:
```diff
@@ -0,0 +1,47 @@
+
+type macAddressValue struct{ v *net.HardwareAddr }
+
+func newMacAddressValue(p *net.HardwareAddr) *macAddressValue {
```
This function isn't exported and only called once. Can it be inlined?

NightTsarina replied:
As you can see, I really don't know how to code in Go :)
I did inline this, and as a result I also simplified the type.

dswarbrick commented on macaddress.go:
```diff
@@ -0,0 +1,47 @@
+}
+
+func (f *macAddressValue) Get() interface{} {
+	return (net.HardwareAddr)(*f.v)
```
This looks redundant.
```suggestion
	return *f.v
```

NightTsarina replied:
I guess I left those casts because I was copying an example, or because a previous version of Go required them. Removed

dswarbrick commented on macaddress.go:
```diff
@@ -0,0 +1,47 @@
+}
+
+func (f *macAddressValue) Set(s string) error {
+	if s == "broadcast" || s == "all" {
+		s = "ff:ff:ff:ff:ff:ff"
+	}
+	if s == "local" {
+		s = "00:b0:52:00:00:01"
+	}
```
I would consider making these special addresses net.HardwareAddr pseudo-constants, e.g.
```go
var (
	bcastMAC = net.HardwareAddr{0xff, 0xff, 0xff, 0xff, 0xff, 0xff}
	localMAC = net.HardwareAddr{0x00, 0xb0, 0x52, 0x00, 0x00, 0x01}
)
```
... and only call `net.ParseMAC` if not one of the predefined aliases.

NightTsarina replied:
done

NightTsarina committed:
Improve and simplify MAC address parser

* Apply code review suggestions:
  - Define constants for broadcast and local addresses.
  - Inline constructor.
  - Better handle parser errors.
* Replace struct with type alias for macAddressValue.
* Avoid unneeded copies and conversions.

NightTsarina commented:
Applied suggestions, and a couple more improvements. I think it looks better now, PTAL?

dswarbrick commented on homeplug_exporter.go:
```diff
@@ -131,6 +133,7 @@ func (n *HomeplugNetworkInfo) UnmarshalBinary(b []byte) error {
 			return err
 		}
 		n.Networks = append(n.Networks, ns)
+		level.Debug(logger).Log("msg", fmt.Sprintf("Network found: %s", &ns))
```
For structured (i.e., machine-readable) logs, this works better:
```suggestion
		level.Debug(logger).Log("msg", "Network found", "network", &ns)
```

NightTsarina replied:
ahh, I wondered sometimes why logs were formatted like this

dswarbrick commented on homeplug_exporter.go:
```diff
@@ -143,26 +146,34 @@ func (n *HomeplugNetworkInfo) UnmarshalBinary(b []byte) error {
 			return err
 		}
 		n.Stations = append(n.Stations, ss)
+		level.Debug(logger).Log("msg", fmt.Sprintf("Station found: %s", &ss))
```
```suggestion
		level.Debug(logger).Log("msg", "Station found", "station", &ss)
```

dswarbrick commented on homeplug_exporter.go:
```diff
@@ -262,13 +279,11 @@ func main() {
 	prometheus.MustRegister(version.NewCollector("homeplug_exporter"))
 
-	level.Info(logger).Log("msg", fmt.Sprintf("Collecting from MAC address %s via interface %s", dest.String(), iface.Name))
+	level.Info(logger).Log("msg", fmt.Sprintf("Collecting from MAC address %s via interface %s", destAddress.String(), iface.Name))
```
The `%s` verb will call `String()` implicitly (if the operand implements the fmt.Stringer interface).
```suggestion
	level.Info(logger).Log("msg", fmt.Sprintf("Collecting from MAC address %s via interface %s", destAddress, iface.Name))
```

dswarbrick commented on macaddress.go:
```diff
@@ -0,0 +1,47 @@
+		if err != nil {
+			return err
+		} else if len(v) != 6 {
+			return errors.New("Invalid address length")
```
Nit: most Go style guides recommend starting error strings with lower case, since they are often cited within log messages.
```suggestion
			return errors.New("invalid address length")
```

NightTsarina replied:
good to know!

dswarbrick reviewed (commented):
Looking pretty good, almost there.

NightTsarina committed:
Apply suggestions from code review

Co-authored-by: Daniel Swarbrick <daniel.swarbrick@gmail.com>

NightTsarina commented:
Thanks for all the very useful comments! PTAL?

dswarbrick commented:
> Thanks for all the very useful comments! PTAL?

LGTM.

NightTsarina commented:
@dswarbrick github blocks merging because the PR has not been approved, could you do that? (I mean, i can force it myself, but I am trying to be tidy and consistent in all my repos :) )
